### PR TITLE
Introduce Money column type #836

### PR DESCRIPTION
**Pull request by encircled**
Adds support for Money columns based on Java Currency and BigDecimal for amounts #836

### CODE REVIEW

**Comment by Tapac** [comment]
@encircled , thank you for a PR, but could you concider to use [JSR 354: Money and Currency API](https://github.com/JavaMoney/jsr354-api) instead of your Money class? It will allow to use other functions from JSR.

**Comment by encircled** [comment]
Sure, make sense

**Comment by Tapac** [comment]
Few more things:
1. Let's move tests to all-tests module and just depend it on your `exposed-money` module. 
2. What do you think about to add possibility to define composite amount + currency field on entity? It will allow to write queries like 'select * from Foo where currency = 'USD'` .
For that purpose you have to create separate `CurrencyColumnType` and also some wrapper/delegate class which will contains two columns and read/write to them on get/set with a Money type.
If you need more details or help on that - feel free to ask.

**Comment by encircled** [comment]
1. Alright, I'll move it to `exposed-tests`. Just wonder what is the reason to do so?

2. I was thinking about it from the beginning, but composite column type is not supported yet, right? I will prepare a draft to check that we see it the same way :)

**Comment by encircled** [comment]
I have a working PoC, entities API looks like:
```java
class AccountDao(id: EntityID<Int>) : IntEntity(id) {
    
    val money : MonetaryAmount by Account.composite_money

    // Optionally it can be mapped as a separated fields on the DAO as well
    val currency : CurrencyUnit by Account.composite_money.currency
    val amount : BigDecimal by Account.composite_money.amount

    companion object : EntityClass<Int, AccountDao>(Account)

}

object Account : IntIdTable("Account") {

    val composite_money = compositeMoney("composite_money", 10)

}
```

And the query allows to:
```java
AccountDao.find {
                Account.composite_money eq Money.of(10, "USD") or
                        (Account.composite_money.currency eq Monetary.getCurrency("USD")) or
                        (Account.composite_money.amount eq BigDecimal.ZERO)

            }
```

**Comment by Tapac** [comment]
Look nice to me, but please support `compositeMoney` column declaration with two different column names for amount and currency to allow map it to existing tables.

**Comment by encircled** [comment]
OK

**Comment by encircled** [comment]
I've left a few `TODO to be discussed`, please check it out

**Comment by Tapac** [comment]
@encircled, I don't think what `CompositeColumn` should be inherited from a `Column` also I don't like the CompositeColumnType.
I make some code improvements but the code isn't in working state for sure.
Could you apply [the patch](https://gist.github.com/Tapac/a5dfa1f37b4a584898dae3e123b88500) and tell me what do you think about it?

**Comment by encircled** [comment]
Yeah, I did not like that inheritance too. Will take a look

**Comment by encircled** [comment]
@Tapac so it was pretty easy to implement without inheritance and it works now. The only question from my side is complicated reusing of extension functions. I've created functions with receiver to do so: Entity#doWithEntity and Table#doWithTable. Alternatively, it is possible to extract current extension functions to avoid this hack. What do you think?

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/CompositeColumn.kt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * @author Vladislav Kisel
+ */
+abstract class CompositeColumn<T> : Expression<T>() {
+
+    /**
+     * Parse values from [compositeValue] and return list of real columns with its values
+     *
+     * @return key - real column, value - its parsed value
+     */
+    abstract fun getRealColumnsWithVales(compositeValue : T) : Map<Column<*>, Any?>
```
Comment (educhastenier) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/CompositeColumn.kt:15]: Did you mean
```suggestion
    abstract fun getRealColumnsWithValues(compositeValue : T) : Map<Column<*>, Any?>
```
?

Comment (encircled) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/CompositeColumn.kt]: Yeah

**Comment by Tapac** [comment]
@educhastenier , about `Entity#doWithEntity` and `Table#doWithTable`. Why not to use simple `with` or `apply` functions?

**Pull request merged**